[Structural] adding prebuckling solver

manuelmessmer opened the pull request:
The Prebuckling Solver computes the critical load multiplier for a given load set at which the structure will buckle. It always refers to the initial, user-defined load. The implementation does not as usually compute the "classical buckling eigenvalue problem" given as (K_mat+ lambda*K_geo)*phi = 0, but relies only on the total stiffness matrix of two consecutive load steps. (K_current + lambda*K_dot)*phi = 0; with K_dot = (( K_next(lambda + h) - K_current) ) / h). Where h is a small change in the load factor. 
Therefore it is not required to compute K_mat and K_geo separately, which would require major changes in the current element implementations. 
To follow the entire prebuckling load path the simulation is conducted iteratively, while the applied loads are modified towards the computed buckling load. It is differentiated between a "small" load step and a "big" load step. The underlying theory of the approximation of K_dot requires a small change in the load factor (small step). After the small step (small change in the load factor) we analyse the eigenvalue problem and find a new load factor. During the big step we push the loads closer to the actual buckling load e.g. to half the value of the computed buckling load. Then another small step and eigenvalue analysis follows etc..
This procedure is repeated until the load factors finally converge. 
In case one wants to compute the linear/theoretical buckling load, the simulation can be stopped after the first small load step. 
The Solver only converges when the initial load is smaller than the actual buckling load. In general it is recommend to apply a very small load.

manuelmessmer committed:
Adding Prebuckling Solver to Structural Mechanics Application.

manuelmessmer committed:
Adding "load_Multiplier" to the eigenvalue post-processing options. "load_multiplier" is used in the structural mechanics prebuckling analysis.

manuelmessmer committed:
Adding Unittest for Structural Prebuckling Analysis

manuelmessmer committed:
Add prebuckling test to structural mechanics test suite.

manuelmessmer committed:
Minor Changes is Prebuckling Strategy

philbucher reviewed (changes requested):
cool addition :+1: 

blocking to avoid accidental merge, I want to review this properly

loumalouomega commented on applications/StructuralMechanicsApplication/custom_python/add_custom_strategies_to_python.cpp:
```diff
@@ -24,6 +24,7 @@
 #include "custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp"
 #include "custom_strategies/custom_strategies/formfinding_updated_reference_strategy.hpp"
 #include "custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp"
+#include "custom_strategies/custom_strategies/prebuckling_strategy.hpp"
```
We are supposed toconsider.h instead of .hpp

philbucher replied:
afaik we didn't reach consensus
I personally prefer `hpp` in this case because header-only and also consistent with the other files

loumalouomega replied:
(I prefer *.hpp too, but it is too late..., the standard is .h...)

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+    ///@name Operations
+    ///@{
+
+    void SetIsInitialized(bool val)
```
Add Doxygen description to everything please

manuelmessmer replied:
Added Doxygen Descriptions.

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        KRATOS_INFO_IF("PrebucklingStrategy", BaseType::GetEchoLevel() > 2 && rank == 0)
+            << "Entering Initialize" << std::endl;
+
+        if (mInitializeWasPerformed == false)
```
```suggestion
        if (!mInitializeWasPerformed)
```

manuelmessmer replied:
Changed

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        {
+            SchemePointerType &pScheme = this->pGetScheme();
+
+            if (pScheme->SchemeIsInitialized() == false)
```
Idem I prefer ! instead of == false
```suggestion
            if (!pScheme->SchemeIsInitialized())
```

manuelmessmer replied:
Changed

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            if (pScheme->SchemeIsInitialized() == false)
+                pScheme->Initialize(rModelPart);
+
+            if (pScheme->ElementsAreInitialized() == false)
```
```suggestion
            if (!pScheme->ElementsAreInitialized())
```

manuelmessmer replied:
Changes

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            if (pScheme->ElementsAreInitialized() == false)
+                pScheme->InitializeElements(rModelPart);
+
+            if (pScheme->ConditionsAreInitialized() == false)
```
```suggestion
            if (!pScheme->ConditionsAreInitialized())
```

manuelmessmer replied:
Changed

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            if (pScheme->ConditionsAreInitialized() == false)
+                pScheme->InitializeConditions(rModelPart);
+        }
+        //initialisation of the convergence criteria
```
```suggestion
        // Initialisation of the convergence criteria
```

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        BuilderAndSolverPointerType &pEigenSolver = this->pGetEigenSolver();
+        pEigenSolver->GetLinearSystemSolver()->Clear();
+
+        // if (this->pGetMassMatrix() != nullptr)
```
Remove if unused

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            BuilderAndSolverPointerType &pEigenSolver = this->pGetEigenSolver();
+            BuilderAndSolverPointerType &pBuilderAndSolver = this->pGetBuilderAndSolver();
+            // Scheme
+            SchemePointerType &pScheme = this->pGetScheme();
```
Pointers does not require &, besides I prefer the & in the side of the type. That's only my opinion but I think is in the style guide

loumalouomega replied:
@philbucher

manuelmessmer replied:
If I remove the & for the Member Variables like StiffnessMatrix, I get an segmentation fault.

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+        typename ConvergenceCriteriaType::Pointer pConvergenceCriteria = mpConvergenceCriteria;
+
+        //initializing the parameters of the Newton-Raphson cycle
```
Please start all comments with space and capital, looks more professional

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        pConvergenceCriteria->FinalizeNonLinearIteration(rModelPart, pBuilderAndSolver->GetDofSet(), rStiffnessMatrix, rDx, rRHS);
+                   
+        if (is_converged)
+        {
```
In the style guide the { is in the declaration

manuelmessmer replied:
Sry, I didn't see it in the style guide. That means in every scope (class, function, if statement etc. ) the "{" needs to be one line up in the declaration?

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+            if (is_converged)
+            {
+                // if (mpConvergenceCriteria->GetActualizeRHSflag())
```
Remove if unused

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            //End Method 2#########################################################################
+
+            // Convergence criteria for buckling analysis
+            if( abs(lambda/lambdaPrev) < mConvergenceRatio )
```
![imagen](https://user-images.githubusercontent.com/19991680/35767644-8a9c6856-08f0-11e8-9944-e72c6c1db7b3.png)

Use std::abs, otherwise will take an integer, otherwise will converge always that is minor of 1.0
```suggestion
            if( std::abs(lambda/lambdaPrev) < mConvergenceRatio )
```

manuelmessmer replied:
abs changed

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+            //End Method 2#########################################################################
+
+            // Convergence criteria for buckling analysis
```
BTW, in my opinion should be in a independent convergence criteria

philbucher replied:
just for one line?

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+    }
+
+    /**
+     * Function to perform expensive checks.
```
```suggestion
     * @brief Function to perform expensive checks.
```

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+    /**
+     * Function to perform expensive checks.
+     * It is designed to be called ONCE to verify that the input is correct.
```
```suggestion
     * @details It is designed to be called ONCE to verify that the input is correct.
```

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+    BuilderAndSolverPointerType mpBuilderAndSolver;
+
+    // SparseMatrixPointerType mpMassMatrix;
```
Remove if unused

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            }
+        } 
+    }
+/*
```
Wrong spaces and missing *
```suggestion
/**
```

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+            // Check if condition is surface load
+            if (it_condition->Has(SURFACE_LOAD))
+            {
+                array_1d<double, 3>& surface_load = it_condition->GetValue(SURFACE_LOAD);
```
style guide->r_surface_load

loumalouomega replied:
```suggestion
                array_1d<double, 3>& r_surface_load = it_condition->GetValue(SURFACE_LOAD);
```

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+/*
+Update Load Conditions
+*/
+    void UpdateLoadConditions()
```
I am against this, I think is more intelligent to define a Variable like: buckling coefficient. (BUCKLING_COEFFICIENT) and define as a property, then check this variable in all load conditions, otherwise take 1. Therefore you will need to change the variable only once. Additionally some people apply loads in the nodes, so this will not work

manuelmessmer replied:
What means "check this in all load conditions"? But I still have to loop over all conditions, right? So it is the same function with a different variable?

armingeiser replied:
That means in every load condition, the multiplication with a Variable (i suggest) `LOAD_FACTOR` should be done when returning the RHS. 

Like this you don't need to store initial loads and change them gradually but you can simply modify the load factor.
However this involves modification of **all** existing conditions.

FYI @KlausBSautter @MFusseder

loumalouomega replied:
@RiccardoRossi what do you think about my proposal?

philbucher replied:
maybe alternatively a scheme could me created that does this multiplication
=> when calculating the RHS of a condition

loumalouomega commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+                for (std::size_t j = 0; j < NumNodeDofs; j++)
+                {
+                    auto itDof = std::begin(NodeDofs) + j;
+                    if( !((*itDof)->IsFixed()) )
```
This does not respect the bracket style guide

loumalouomega reviewed (changes requested):
Several comments and a big error (abs->std::abs)

loumalouomega commented:
Travis does not compile

philbucher commented on applications/StructuralMechanicsApplication/custom_python/add_custom_strategies_to_python.cpp:
```diff
@@ -94,6 +96,13 @@ void  AddCustomStrategiesToPython(pybind11::module& m)
+    // Prebuckling Strategy
+    py::class_< PrebucklingStrategyType, typename PrebucklingStrategyType::Pointer,BaseSolvingStrategyType >(m,"PrebucklingStrategy")
+        .def(py::init<ModelPart&, BaseSchemeType::Pointer, BuilderAndSolverPointer, BuilderAndSolverPointer, ConvergenceCriteriaPointer, int, double, double, double, double >(), 
+            py::arg("model_part"), py::arg("scheme"), py::arg("builder_and_solver"), py::arg("static_solver"), py::arg("convergence_criteria"), py::arg("max_iteration"), py::arg("initial_step"), py::arg("small_step"), py::arg("big_step"), py::arg("convergence_ratio") )
```
please remove the `py::arg`, we typically don't use it

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_analysis.py:
```diff
@@ -0,0 +1,69 @@
+# Importing the base class
+from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis
+
+class StructuralMechanicsPrebucklingAnalysis(StructuralMechanicsAnalysis):
```
can you explain why you need a custom Analysis?

I would really prefer to avoid this

manuelmessmer replied:
I derived the analysis stage, because I don't know in advance how many load steps the solution will need.  I basically use the "end_time" as maximal load iteration.
So the RunSolutionLoop is overriden to break, when the solution is converged. Are there better options to do this?
Futhermore, I only want to postprocess the eigenvalues in every second load step, because the eigenvalue problem is only solved every second step.

manuelmessmer replied:
@philbucher is this ok?

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_solver.py:
```diff
@@ -0,0 +1,114 @@
+        self.SetSolverFlag( "eigen")
+        eigen_solver = self.get_builder_and_solver() # The eigensolver is created here.
+        self.SetSolverFlag( "static")
+        builder_and_solver = self.get_builder_and_solver() # The linear solver is created here.
```
Please restructure this, it should follow the same approach as in thh baseclass
i.e you could have separate fcts for `get_builder_and_solver_eigen` and `get_builder_and_solver_static` etc

manuelmessmer replied:
I added the get_builder_and_solver_eigen function and all nested "_eigen" functions, like create_builder_and_solver_eigen etc., as well. For the "normal" builder_and_solver I now just use the functions from the base class (get_builder_and_solver). Hope this is ok.

philbucher commented on applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py:
```diff
@@ -282,6 +284,8 @@ def AssembleTestSuites():
     # Dynamic Eigenvalue Analysis
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TTestDynamicEigenvalueAnalysis]))
+    # Buckling analysis test
+    smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TTestPrebucklingAnalysis]))       
```
the tests in the smallsuite must be very fast, > 0.05s on your machine

otherwise please add it to the nightsuite

manuelmessmer replied:
It is 0,2 on my machine -> so changed it to nightSuite

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        mInitializeWasPerformed = val;
+    }
+
+    bool GetIsInitialized() const
```
are you sure you need all those access-functions? Seems to me like you are not using them. Also arent they already in the baseclass?

manuelmessmer replied:
Removed most of them!

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+        ModelPart &rModelPart = BaseType::GetModelPart();
+        const int rank = rModelPart.GetCommunicator().MyPID();
+        typename ConvergenceCriteriaType::Pointer pConvergenceCriteria = mpConvergenceCriteria;
```
why in this functions you create all those aux-variables

manuelmessmer replied:
I removed them.

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        KRATOS_TRY
+
+        ModelPart &rModelPart = BaseType::GetModelPart();
+        const int rank = rModelPart.GetCommunicator().MyPID();
```
```suggestion

```

this is no longer used, you can remove it from everywhere

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        SparseMatrixType &rStiffnessMatrix = this->GetStiffnessMatrix();
+        SparseMatrixType &rStiffnessMatrixPrevious = this->GetStiffnessMatrixPrevious();
+        SparseVectorType &rDx = this->GetDx();
+        SparseVectorType &rRHS = this->GetRHS();
```
why don't you just access the member pointer directly? Would be much cleaner IMO

philbucher commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        double deltaLoadMultiplier = 0.0;
+        if( mLoadStepIteration == 1) // inital step
+        {
+            deltaLoadMultiplier = mInitialStep*(lambda + lambdaPrev);
```
please follow the style guide for naming of variables

philbucher reviewed (commented):
@manuelmessmer looks not too bad, I made some comments

in the strategy there is still some commented code and also probably some unnecessary temporary variables

manuelmessmer committed:
using std::abs

manuelmessmer committed:
Added Doxygen Description

Deleted unnecessary comments.
Changed beginning of comments to capital letters.

manuelmessmer committed:
Changed if-statement arguments from " == false" to "!.."

manuelmessmer committed:
Changed some variable names.

manuelmessmer committed:
Access member pointers directly

And delete corresponding getter methods

manuelmessmer committed:
Remove unused variables and functions.

Add Doxygon description to get methods.

manuelmessmer committed:
Add test to nightSuite, Remove py::arg

manuelmessmer committed:
Restructured Prebuckling Solver

manuelmessmer committed:
Added two spaces.

armingeiser commented on applications/StructuralMechanicsApplication/custom_processes/postprocess_eigenvalues_process.cpp:
```diff
@@ -181,9 +181,12 @@ std::string PostprocessEigenvaluesProcess::GetLabel(const int NumberOfEigenvalue
     } else if (lable_type == "frequency") {
         label += "_EigenFrequency_[Hz]_";
         label_number = std::sqrt(EigenvalueSolution) / (2 * Globals::Pi);
+    } else if(lable_type == "load_multiplier"){
```
"load_factor" seems more familiar to me (personal preference)

armingeiser commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_analysis.py:
```diff
@@ -0,0 +1,69 @@
+        solver_settings = project_parameters["solver_settings"]
+        super(StructuralMechanicsPrebucklingAnalysis, self).__init__(model, project_parameters)
+    
+    '''Break Solution Loop when Buckling Analysis is converged.
```
comment should be below function definition

armingeiser commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_solver.py:
```diff
@@ -0,0 +1,114 @@
+    def GetDefaultSettings(cls):
+        this_defaults = KratosMultiphysics.Parameters("""{
+            "buckling_settings"     : {
+                "initial_step"          : 1.0,
```
initial/small/big do not describe very well what these numbers are used for. Maybe you can find a better naming.

armingeiser commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+
+/*
+Store Initial Loads to ensure that load multiplier always refers to user input
+*/
```
Indentation of comment does not match function indentation

armingeiser commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+                rNodeEigenvectors.resize(NumEigenvalues, NumNodeDofs, false);
+            }
+
+            // TO BE VERIFIED!! In the current implmentation of Dofs there are nor reordered and only pushec back.
```
reminder

armingeiser commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/prebuckling_strategy.hpp:
```diff
@@ -0,0 +1,980 @@
+        {
+            auto it_condition = it_conditions_begin + j;
+            // Check if condition is point load
+            if (it_condition->Has(POINT_LOAD))
```
Here only `POINT_LOAD` and `SURFACE_LOAD` are considered. @philbucher i am pretty sure there are other loads as well. Like this only certain loads will be considered to be dependent on the load factor.

 I think this is related to @loumalouomega comment:

> I am against this, I think is more intelligent to define a Variable like: buckling coefficient. (BUCKLING_COEFFICIENT) and define as a property, then check this variable in all load conditions, otherwise take 1. Therefore you will need to change the variable only once. Additionally some people apply loads in the nodes, so this will not work

manuelmessmer replied:
Okay, so then we do it like that and I change the implementation of all conditions?

armingeiser replied:
Before you start doing that let's wait for @philbucher or other opinions because thats something that should be discussed first.

philbucher replied:
to me modifying all conditions for this is a no-go

manuelmessmer replied:
Any more suggestions how we wanna proceed?

RiccardoRossi commented:
I am not in favor of modifying all the conditions...

the loads is changed from outside, and in my opinion we should keep it that way, even if that is not the natural choice for load control

RiccardoRossi commented:
what about having a different process for the application of the loads.

somethign like "ApplyFactorProportionalVariable" which we know will apply the applied value multiplied by a factor, taken for example from the process info.

this way we would have to make no change to the conditions, and we could still achieve the same

manuelmessmer committed:
Merge branch 'master' into structural/adding_prebuckling_solver

manuelmessmer commented:
Hey guys, I would like to revive this PR. There was still an open question about how to modify the external loads during the analysis steps. Riccardo suggested to create a new process to do that. However, yesterday we came up with another idea. What if we use the parameter TIME t as load multiplier/ load factor. This solution seems straight forward and also works for any kind of loading. What do you think?

manuelmessmer commented:
ping @philbucher @armingeiser @loumalouomega

loumalouomega reviewed (dismissed):
OK

philbucher commented:
> Hey guys, I would like to revive this PR. There was still an open question about how to modify the external loads during the analysis steps. Riccardo suggested to create a new process to do that. However, yesterday we came up with another idea. What if we use the parameter TIME t as load multiplier/ load factor. This solution seems straight forward and also works for any kind of loading. What do you think?

sorry for the late answer
I don't think it is a good idea hardcoding it to to `TIME`. This restricts the usability very much
Since we usually do this with processes I agree with @RiccardoRossi

RiccardoRossi commented:
Hi @philbucher the proposal of using TIME is mine in this case.

the point is that buckling analysis is used in pseudo time, so in this case it wouldn't matter.

if we don't do this, then we need to define a "factor_proportional_loading" that overwrites the loading field. for example imagine that you want to apply a pressure with a space distribution of 
"x+y" so that this is applied as "(x+y)*F"  where "F" is a factor taken from example from the process info.
To do so we would need to redefine all of the loading processes to parse the "x+y" and then to multiply it by F.

instead we could just write that the load is "(x+y)*t" and it would work out of the box since the pseudo time is adjusted

philbucher commented:
@RiccardoRossi ok, no objections then

manuelmessmer committed:
Merge branch 'master' into structural/adding_prebuckling_solver

manuelmessmer committed:
Make Parameter TIME load_multiplier.

Minor changes of variable naming.

manuelmessmer committed:
Update conditions using TIME.

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py:
```diff
@@ -45,8 +45,11 @@ def CreateSolverByParameters(model, solver_settings, parallelism):
         elif solver_type == "adjoint_static":
             solver_module_name = "structural_mechanics_adjoint_static_solver"
 
+        elif solver_type == "prebuckling_analysis":
```
```suggestion
        elif solver_type == "prebuckling":
```

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_analysis.py:
```diff
@@ -0,0 +1,133 @@
+            self.InitializeSolutionStep()
+            self._GetSolver().Predict()
+            is_converged = self._GetSolver().SolveSolutionStep()
+            self.__CheckIfSolveSolutionStepReturnsAValue(is_converged)
```
I don't think you need this here, this should be working in StructuralMechanics
```suggestion

```

manuelmessmer replied:
You are right. Works without it!

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_analysis.py:
```diff
@@ -0,0 +1,133 @@
+            for process in self._GetListOfProcesses():
+                process.ExecuteAfterOutputStep()
+
+    def __CreateListOfProcesses(self):
```
why do you override this?

manuelmessmer replied:
I need to override Initialize(), where this function is called. But this function is private, so can't be called from the derived class. Am I right?

philbucher replied:
well its python, you can still call them
it is more cumbersome, but you can do it. I don't know the syntax by hard but you can google it

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_analysis.py:
```diff
@@ -0,0 +1,133 @@
+        self._list_of_output_processes = self._CreateProcesses("output_processes", order_processes_initialization)
+        self._list_of_processes.extend(self._list_of_output_processes) # Adding the output processes to the regular processes
+
+    def __CheckIfSolveSolutionStepReturnsAValue(self, is_converged):
```
can be removed, see comment above

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_solver.py:
```diff
@@ -0,0 +1,110 @@
+        return this_defaults
+
+    #### Private functions ####
+    def AdvanceInTime(self, current_time):
```
why overridden?

manuelmessmer replied:
Because the new time depends on the computed buckling load. As the TIME parameter is considered to be the load multiplier.

loumalouomega replied:
I spoke with @RiccardoRossi and I said that I think this is a bad idea. Many processes, utilities, strategies,... expect the time to have an specific meaning, this may lead to many unexpected behaviours and potential problems a the code may broke eventually. I think is better to define the load factor with another variable (for example LAMBDA) and modify the processes that compute the loads in order to allow another variable (for example "l"). In that case you don't need to consider "t" and the desired feature is preserved. Opinion @philbucher ?

RiccardoRossi replied:
For the moment we have no better solution, so i think it should be accepted as it is.
If we eventually add the possibility of having another variable in the input functions to have this kind of behaviour we will eventually modify this, but i really would not do it in this PR (which is complex enough as of now)

loumalouomega replied:
As you see I am not blocking, but I think the idea deserves consideration

manuelmessmer committed:
Review comments!

manuelmessmer committed:
Codacy suggestions.

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_prebuckling_analysis.py:
```diff
@@ -0,0 +1,120 @@
+    def __init__(self, model, project_parameters):
+        super(StructuralMechanicsPrebucklingAnalysis, self).__init__(model, project_parameters)
+
+    def Initialize(self):
```
TBH I still don't see why you need to override it completely
the things you do in here you could IMO also do with your own initialize like this:
~~~py
def Initialize(self):
    super().Initialize()
    if problem_data.Has("start_time"):
    # ... and so on
~~~

manuelmessmer replied:
Well, that makes a lot of sense I am afraid. Will do the changes.

manuelmessmer replied:
Ah now I remember. The super.Initialze() requires start_time and end_time to be defined. However, they are not needed for the buckling solver.

philbucher replied:
hm
maybe then sth like this:
~~~py
def Initialize(self):
    # set dummy start_time and end_time
    super().Initialize()
    # remove dummy start_time and end_time
    if problem_data.Has("start_time"):
    # ... and so on
~~~

this might be a bit hacky but then you could use it a is

I took not that this should be improved in the future in the `AnalysisStage`. After that is done there will be no more need for your hack solution

manuelmessmer replied:
Makes sense. Pushed the changes!

philbucher reviewed (dismissed):
aside from two comments I am ok with it, nice work :+1:

I think there are still a few things that can be improved but I don't want to keep this blocked for longer

BTW can you post an example? I like colorful pictures:

manuelmessmer committed:
Override _create_convergence_criterion

Just noticed that I normally used the displacement_criterion. Makes more sense and converges better than the residual_criterion, as external loads are manipulated during the analysis. Here I make sure it always uses the displacement_criterion.

manuelmessmer committed:
Call private function __CreateListProcessess

manuelmessmer commented:
[Example.zip](https://github.com/KratosMultiphysics/Kratos/files/4274846/Example.zip)
Here is an example of an classical Euler column. Once this branch is merged you can rerun it. But for now I also pushed the Gid results for the eigenvectors and the displacement.

manuelmessmer committed:
Review comment!

philbucher reviewed (approved):
ok :+1: 
ping @armingeiser

RiccardoRossi commented:
MINOR REQUEST: when you merge, do not delete the branch, this will make it easy to update once i change the load function to add the possibility of taking the load multiplier from the process info

manuelmessmer commented:
Perfect. Thank you for the help @philbucher. Btw, as this is part of my master thesis, there will be a presentation (related to this and a bit more) also at the static chair in Munich. For those who are interested, it will take place most likely in the beginning of April.

philbucher commented:
> MINOR REQUEST: when you merge, do not delete the branch, this will make it easy to update once i change the load function to add the possibility of taking the load multiplier from the process info

whats the difference compared doing it from master?

RiccardoRossi commented:
> > MINOR REQUEST: when you merge, do not delete the branch, this will make it easy to update once i change the load function to add the possibility of taking the load multiplier from the process info
> 
> whats the difference compared doing it from master?

just that all the files to change will be colelcted in a single PR, so substitution is easier

armingeiser reviewed (approved):
I don't have time for a detailed review now, so i don't want to block this. My comments seem not to be relevant anymore.

philbucher commented:
@manuelmessmer your test fails in the CI:
~~~
======================================================================
FAIL: test_dynamic_eigenvalue_analysis (test_prebuckling_analysis.TestPrebucklingAnalysis)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/Kratos/Kratos/bin/Release/applications/StructuralMechanicsApplication/tests/test_prebuckling_analysis.py", line 276, in test_dynamic_eigenvalue_analysis
    self._check_load_multiplier(loadFactor_sym,loadFactor_full, reference_value )
  File "/__w/Kratos/Kratos/bin/Release/applications/StructuralMechanicsApplication/tests/test_prebuckling_analysis.py", line 249, in _check_load_multiplier
    self.assertLess( abs(1-load_multiplier1[0]/load_multiplier1[8]), 1.0e-4)
AssertionError: 0.3986418020668909 not less than 0.0001
~~~
can you please take a look?

manuelmessmer commented:
Yes will look at it.

philbucher commented:
thanks

manuelmessmer commented:
@philbucher sry for the late response. However, the problem is not really straightforward. At least not to me. First of all, it only fails for gcc, the clang compilation does not show any error of the test. There is also no error on my machine. Neither in Release nor in FULL Debug. I also enabled -D_GLIBCXX_DEBUG and tried a couple of more gcc flags -> no error. 
Additionally, I asked Riccardo and Raul to run the test - all worked fine. 
Do you have any suggestions?

philbucher commented:
hi @manuelmessmer 
hm this sounds like an issue with tolerances &| shared-memory parallelism (OpenMP)
Have you tried with different number of omp threads?

Also maybe you have some instability/tolerance that is triggered or not, depending on the machine.

manuelmessmer commented:
Yes I tried different number of threads, everything from 1 to 8. 

I also created a another branch where I switched the test to the smallSuite and printed the actual values of the very first iteration (far away from the buckling point). On top of that, I played with different tolerance value. 
The numbers are close but just not the same: 92.57012573773818 != 92.96764052069236
Of course in the next iteration it gets worse and worse.
Comparison my machine: 92.96772626270567 = 92.9677262533037 (exactly same results until the last digit on Raul's machine with again a different amount of threads)

philbucher commented:
hmmm
have you tried running with valgrind?

manuelmessmer commented:
[screen.log](https://github.com/KratosMultiphysics/Kratos/files/4298401/screen.log)

Sry, I am not really familiar with that, I ran it with the following command:
PYTHON_MALLOC=malloc valgrind --leak-check=full --track-origins=yes  python3 test_prebuckling_analysis.py 
I also posted the log file.

philbucher commented:
hm looks clean
what linear solver are you using?

FYI I will be on vacation next week and will probably not answer :)

manuelmessmer commented:
Standard: SkylineLUFactorizationSolver
The Eigen library is now inside the core right? So there is no chance that there is another version used in the CI?

All right. Thanks for the info. Enjoy your holiday ;)

philbucher commented:
> The Eigen library is now inside the core right? So there is no chance that there is another version used in the CI?

Its not in the core but in the EigenSolversApplication. Hence you only need to set it to on in the configure to use it (which I recommend, much faster)

philbucher commented:
> So there is no chance that there is another version used in the CI?

You specify the solver explicitly, hence no other Solver can be used

BTW why did you set up the problem for the test by hand? I.e. why not through json? I have the impression that esp for complex things it is easy to miss sth that would be taken care of automatically if done through the regular workflow
Do you have the json setup for this case?

manuelmessmer commented:
There is no particular reason. But I can try to push the same test with .json to my dummy branch.

manuelmessmer commented:
I set up the same test using the analysis stage. As expected same problem. I put it here #6519.

RiccardoRossi commented:
manuel, the problem here is most probably in the calculation of the eigenvalues, which is an iterative process. take a look if the parameters allow you specifying a stricter tolerance

manuelmessmer commented:
I tried to trace the error back. It is not a problem of the eigensolver.
Actually, I printed the L2norm of both matrices given to the eigensolver and the L2norm of the one in the cluster is exactly 2 times larger (until the last digit) than the one on my machine. To be a bit more accurate: 
I conduct 2 loadsteps. In the first loadstep the matrices (K_tan1) differ exactly by the factor 2.0.
In the second load step (K_tan2) they differ by the factor 2.0000001202922078. 
Then, I compute the difference between (K_tan2 - K_tan1) before it is given to the eigenvalue solver with the form: eig(K_tan1, K_tan2-K_tan1).
Those matrices differ again exactly by the factor 2.0.

**Ouput CI:**
After Build and Solve:
NormLoadStep1:  `13689658.759508904` (K_tan1)
NormLoadStep2:  `13689657.936129315` (K_tan2) (very minor change in the force)
Just before eigensolver:
Norm1:  `13689658.759508904` (K_tan1, same as before)
Norm2:  `8.876618804367528` (K_tan2 - K_tan1)

**Output local:**
After Build and Solve:
NormStep1:  `27379317.51901781` (K_tan1)
NormStep2:  `27379315.872258626` (K_tan2)
Just before eigensolver:
Norm1:  `27379317.51901781` (K_tan1, same as before)
Norm2:  `17.753237608735056` (K_tan2 - K_tan1)

philbucher commented:
I am a bit confused as to what this means, where do you think the problem is now?

manuelmessmer commented:
Well now I am even a bit more confused. When I run the test on my machine on only 2 threads then I obtain the same norm as on the cluster.  Is the function TSparseSpace::TwoNorm() thread dependent?
For each number of thread I get a different norm (everything else stays the same, only output of norm is different). Anyway, with 2 threads I get exactly the same norms as on the CI. So matrices before eigensolver seem to be the same, however eigenvalues are not (not perfectly).   Sry for the confusion.

manuelmessmer commented:
I try to recap again. I guess the function TSparseSpace::TwoNorm() didn't do what i expected it to do.  For every number of thread I get a different result of the norm. Nevertheless, the eigenvalues and the overall results of the solver stay the same. Therefore, I think there is a problem with the paralellization of TSparseSpace::TwoNorm() . Neverthelesse, with 2 threads I get exactly the same numbers as on the CI (until the last digit). The norms are 100% the same. Therefore, I assume the matrices I give to the eigensolver do look the same. However, the eigenvalues are not the same.

philbucher commented:
hm this sounds strange
I checked the implementation ot `TwoNorm`, looks correct to me
@manuelmessmer can you please comment [this line](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/spaces/ublas_space.h#L292) and report the results?
This disables omp for the `TwoNorm`

manuelmessmer commented:
Will do ASAP. I am trying to come back to Germany right now ;) Will do it tomorrow.

philbucher commented:
good luck!

manuelmessmer commented:
Ok when removing that line, the norm is constant for a varying number of threads on my local machine. I just pushed the changes to #6519 to check the results in the CI.
In the TwoNorm implementation, shouldn't it be "#pragma **for** reduction(...)"

manuelmessmer commented:
Well now things make much more sense.
Output CI:
Norm1:  9680050.540978566 (matrix1)
Norm2:  6.276717350576303 (matrix2)
Eigenvalue:  91.57012573774043
Output local:
Norm1:  9680050.540978566 (matrix1)
Norm2:  6.276717350576303 (matrix2)
Eigenvalue:  91.96772626337587
That means the matrices are identical. Nevertheless, the eigensolver returns a slightly different eigenvalue. Could this be related to the MKL support in the CI?

philbucher commented:
Hi @manuelmessmer 
I just figured out together with @RiccardoRossi that the `TwoNorm` function is completely wrong, I am working on a fix, hence I would not take trust its results too much atm

regarding MKL, yes this is possible. In our experiences the differences are very small and usually negligible. 
You could install the MKL locally and check if it makes a big difference for you. Check the instructions [here](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/EigenSolversApplication#install-mkl-on-ubuntu-with-apt)

manuelmessmer commented:
Hi @philbucher and @RiccardoRossi,
we are getting closer ;)
The discrepancy is really introduced by MKL. 
Eigenvalues MKL enabled: 91.57012573773879 (which is pretty much the solution in the CI)
Eigenvalue MKL disabled:  91.96772626337587

philbucher commented:
hm strange 
as I said this is new to me, usually the results are basically identical
@armingeiser @oberbichler have you seen this behavior before?

@manuelmessmer do you get the same results when you decrease the tolerance?

@RiccardoRossi thought?

armingeiser commented:
No, but i am also not using the EigenSystemSolver frequently.

I think we need to figure out where the difference is coming from. The direct solution (with pardiso_ldlt or eigen_lu)  inside the EigenSystemSolver, or the iterative process itself. 

* switching to paridso_lu instead of pardiso_ldlt could be one first step.
 https://github.com/KratosMultiphysics/Kratos/blob/c323ab4dd3c7895550b61d2aa3d9c6932d2add18/applications/EigenSolversApplication/custom_solvers/eigensystem_solver.h#L180

* looking at the result of the solution of the direct solver in the first inner iteration as well (it is nc solution vectors per iteration - see inner loop)
https://github.com/KratosMultiphysics/Kratos/blob/c323ab4dd3c7895550b61d2aa3d9c6932d2add18/applications/EigenSolversApplication/custom_solvers/eigensystem_solver.h#L198

@manuelmessmer how many iterations are needed and what is the tolerance you are using?

manuelmessmer commented:
The tolerance I was using to get the numbers above is tol=1e-10 and I only compute one eigenvalue. Before that, I defined a tolerance tol=1e-6 (This is already more than enough to pass the test without MKL). The difference between the given tolerances with MKL enabled wasn't large. 
I don't remember the required number of iterations. I am running something in a different branch right now. But let you know later.

oberbichler commented:
I don't know what you do exactly but when the condition number of the matrices is really low, MKL can compute slightly different results.

manuelmessmer commented:
@oberbichler The first matrix is just the stiffness matrix of the problem. However, the second matrix is the difference of the stiffness matrix among two load steps. I guess the condition number is similar. However, the values are much much smaller, as the buckling theory requires a very small load increment between the two consecutive load steps.
A look at the norms shows how different they are. Could this be a problem for MKL?
Norm1: 9680050.540978566 (matrix1)
Norm2: 6.276717350576303 (matrix2)

oberbichler commented:
Important is the condition number of the matrix of the linear equation system you are solving with MKL. I am not using Kratos but I observed in another project that Pardiso provides slightly different results then SparseLU for ill-conditioned matrices.

When I wrote the EigenSystemSolver I have done some benchmarks and compared the results with commercial tools. But of course there still might be a bug.

Have you tried the suggestions from @armingeiser ?

In addition I would suggest that you extract the eigenvalue problem (e.g. store it as .mtx). Then you can use other tools to compute a reference solution for validation.

manuelmessmer commented:
If I switch to EigenPardisoLUSolver as suggested by @armingeiser  the problem does not occur. The eigenvalues are the same as with the EigenSparseLUSolver.
So the eigenvalues only deviate for the EigenPardisoLDLTSolver<double>.
Here again the numbers:
EigenSparseLUSolver: 91.96772626337587
EigenPardisoLUSolver: 91.9677262633749
EigenPardisoLDLTSolver: 91.57012573773879

I am pretty sure that the first two eigenvalues are correct. In the test I compare two models. One with symmetry conditions (quarter of the plate) and one without symmetry conditions (full plate). For both LUSolvers the eigenvalues for both models are the same with a very very high accuracy (as they should be).

RiccardoRossi commented:
mmm,

arr u sure your matrix is symmetric?

On Fri, Mar 20, 2020, 2:27 PM manuelmessmer <notifications@github.com>
wrote:

> If I switch to EigenPardisoLUSolver as suggested by @armingeiser
> <https://github.com/armingeiser> the problem does not occur. The
> eigenvalues are the same as with the EigenSparseLUSolver.
> So the eigenvalues only deviate for the EigenPardisoLDLTSolver.
> Here again the numbers:
> EigenSparseLUSolver: 91.96772626337587
> EigenPardisoLUSolver: 91.9677262633749
> EigenPardisoLDLTSolver: 91.57012573773879
>
> I am pretty sure that the first two eigenvalues are correct. In the test I
> compare two models. One with symmetry conditions (quarter of the plate) and
> one without symmetry conditions (full plate). For both LUSolvers the
> eigenvalues for both models are the same with a very very high accuracy (as
> they should be).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5938#issuecomment-601700142>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEJ6JODRJNBYXDSHOXTRINVM7ANCNFSM4JN4MWEA>
> .
>

armingeiser commented:
Both results are "correct" just slightly different because of numerical tolerances.

LU, LLT and LDLT are different methods to solve a linear system, and even though all solvers you used here are direct solvers, they accuracy is dependent on the e.g. condition number of the problem. Here this effect might even be amplified by the fact that this dierct solution happens in a loop inside an iterative process...

I can not find the table anymore, but i think i remember that LDLT is the most sensitive one.

@philbucher @oberbichler I think the same way as we use pardiso_lu as default (over llt/ldlt) for the linear solver in StructMesh we could also use pardiso_lu as default inside the EigensystemSolver.

armingeiser commented:
> arr u sure your matrix is symmetric?

Would of course explain it... But it should be, right? The difference of two tangent stiffness matrices should still be symmetric. 

@oberbichler do you remember if pardiso llt/ldlt fail if called on a non symmetric matrix?

manuelmessmer commented:
i doubled checked it. Yes both matrices are symmetric and have the same dimensions.

RiccardoRossi commented:
Well, in theory LDLt should work for any SPD matrix (and of course deliver the same result as LU since it is unique). There might be some instability problem for very badly conditioned matrices, but i hope this is not the case for you. Also there might be instabilities if you have an entry of D close to zero ... 

in any case, sorry for insisting, but did u try doing

scipy.sparse.linalg(A - A.transpose())/scipy.linalg.norm(A) ?

as a final comment, try by amgcl (with ruge_stuben coarsening) it should also work nicely in this context

manuelmessmer commented:
Ok now checked both matrices.  I am doing this. Should be equivalent to what you suggested, correct?
`            rStiffnessMatrix = rStiffnessMatrixPrevious - rStiffnessMatrix;

            norm1 =  TSparseSpace::TwoNorm(rStiffnessMatrixPrevious);
            norm2 =  TSparseSpace::TwoNorm(rStiffnessMatrix);
            double sum = 0.0;
            for( int i = 0; i < rStiffnessMatrix.size1(); i++)
            {
                for( int j = 0; j < rStiffnessMatrix.size2();j++)
                {
                    sum += std::pow(rStiffnessMatrix(i,j) - rStiffnessMatrix(j,i),2);
                }
            }
            KRATOS_WATCH( sqrt(sum)/norm2 );`
Matrix1: sum/norm1 : 1.16671e-08
Matrix2: sum/norm2 : 0.0535411
The first one is the unmodified stiffness matrix. The second one of the difference of the two loadsteps. Seems like the sustraction of to very similar numbers introduces the problem right.

manuelmessmer commented:
Sry, I dind't do the normalization before. The both values are very small.

manuelmessmer commented:
I also extracted the matrices and gave it to python.
Already the unmodified stiffness matrices reveal a unsymmetry.
scipy.linalg.norm(A - A.transpose())/scipy.linalg.norm(A) gives:
matrix1 = 3.47670469772e-08
matrix2 = 6.91984286814e-08
For the difference it is even larger:
matrix2-matrix1 = 0.0535411131318

In fact scipy.eig() is not even able to solve the eigenvalue problem. But if I first create a symmetric matrix1 and symmetric matrix2 with:
A = 0.5 * (A + A.transpose())
The the first eigenvalue is: 92.92736428286015

manuelmessmer commented:
Moreover, I did the same thing in my strategy in Kratos.
A = 0.5 * (A + A.transpose())
-> The test does not fail anymore, even with MKL and EigenPardisoLDLTSolver.

oberbichler commented:
Assuming that the assembling in Kratos has no bug it might be an error in one of the elements. Or there is some special „unsymetric“ element involved.
I dont remember if the EigenSystemSolver would also work with unsymetric matrices but I think it does not.

oberbichler commented:
Maybe you should also check again the operatioms to compute Matrix2

RiccardoRossi commented:
one last question: ifnyou run with 1 thread, donthebresykts change?
(@maceli)

On Fri, Mar 20, 2020, 5:46 PM manuelmessmer <notifications@github.com>
wrote:

> I also extracted the matrices and gave it to python.
> Already the unmodified stiffness matrices reveal a unsymmetry.
> scipy.linalg.norm(A - A.transpose())/scipy.linalg.norm(A) gives:
> matrix1 = 3.47670469772e-08
> matrix2 = 6.91984286814e-08
> For the difference it is even larger:
> matrix2-matrix1 = 0.0535411131318
>
> In fact scipy.eig() is not even able to solve the eigenvalue problem. But
> if I first create a symmetric matrix1 and symmetric matrix2 with:
> A = 0.5 * (A + A.transpose())
> The the first eigenvalue is: 92.92736428286015
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5938#issuecomment-601800470>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEJ7Y64SCRWRTTWKZITRIOMUXANCNFSM4JN4MWEA>
> .
>

RiccardoRossi commented:
I think that in any case the  symmetrization fix should be committed.

and we need to look into the elements to understand if the source on lack of symmetry comes from them

manuelmessmer commented:
Yes it does come from there. I printed the Element Stiffness Matrix and it is not symmetric. To be more precise, the unsymmetry is introduced when mpCoordinateTransformation->FinalizeCalculations() is called for the coordinate transformation. Here you see the "symmetry norm" before and after that function call. I also put one full stiffness matrix in the text file.
[ElementMatrix.txt](https://github.com/KratosMultiphysics/Kratos/files/4362866/ElementMatrix.txt)

Before: 4.18142e-17
After: 5.99537e-06
Before: 4.18142e-17
After: 6.89148e-06
Before: 4.18142e-17
After: 6.47588e-06
Before: 4.18142e-17
After: 6.48332e-06
Before: 4.18142e-17
After: 6.89156e-06
Before: 4.18142e-17
After: 1.04424e-05

oberbichler commented:
I remember that the shell elements in Kratos have asymmetric stiffness matrices until they are converged(?)

The plot `Abs(K-K^T)` for the matrix in the text file:
![image](https://user-images.githubusercontent.com/7000594/77226834-8f723f00-6b7b-11ea-94e5-e872e219dbe4.png)

`Max(Abs(K)) ~ 430 000`

armingeiser commented:
> I dont remember if the EigenSystemSolver would also work with unsymetric matrices but I think it does not.

https://github.com/KratosMultiphysics/Kratos/blob/d2413c6cfa594179a793df52cabbaa0f3930c6d4/applications/EigenSolversApplication/custom_solvers/eigensystem_solver.h#L85

I think we need to switch the linear solver to pardiso_lu, because the matrix K (matrix 1) - the matrix that is used in the linear solution -  does not have to be positive definite, however the LLT and LDLT can only solve SPD matrices.

oberbichler commented:
I used the LDLT because for a viable structure I would assume the stiffness matrix to be SPD. Might be different in this use case.

Are the tests failing because of a non-positive definite stiffness matrix or because of an incorrect stiffness matrix?

In the first case we definitely should switch to LU.

manuelmessmer commented:
Ok, then I switch to pardiso_lu. 
@RiccardoRossi and @philbucher do you want me to push the changes here?

philbucher commented:
Can you please make a separate PR?

armingeiser commented:
Switching to LU will solve the problem with the failing test here. But the question why the stiffness matrix is not symmetric and if that is acceptable is still open.

manuelmessmer commented:
> I remember that the shell elements in Kratos have asymmetric stiffness matrices until they are converged(?)
> 
> The plot `Abs(K-K^T)` for the matrix in the text file:
> ![image](https://user-images.githubusercontent.com/7000594/77226834-8f723f00-6b7b-11ea-94e5-e872e219dbe4.png)
> 
> `Max(Abs(K)) ~ 430 000`

Maybe I got this wrong, what does it mean that the stiffness matrix of the shell elements are asymmetric until they are converged?

RiccardoRossi commented:
if you look element by element, the LHS is also asymmetric elementwise?

On Thu, Mar 26, 2020 at 10:39 AM manuelmessmer <notifications@github.com>
wrote:

> I remember that the shell elements in Kratos have asymmetric stiffness
> matrices until they are converged(?)
>
> The plot Abs(K-K^T) for the matrix in the text file:
> [image: image]
> <https://user-images.githubusercontent.com/7000594/77226834-8f723f00-6b7b-11ea-94e5-e872e219dbe4.png>
>
> Max(Abs(K)) ~ 430 000
>
> Maybe I got this wrong, what does it mean that the stiffness matrix of the
> shell elements are asymmetric until they are converged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5938#issuecomment-604326945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEIVNREWKQEICJO5Z7DRJMPEXANCNFSM4JN4MWEA>
> .
>


-- 


*Riccardo Rossi*

PhD, Civil Engineer


member of the Kratos Team: www.cimne.com/kratos

Associate Professor at Universitat Politècnica de Catalunya, BarcelonaTech
(UPC)

Full Research Professor at International Center for Numerical Methods in
Engineering (CIMNE)


C/ Gran Capità, s/n, Campus Nord UPC, Building C1, First Floor

08034 – Barcelona – Spain – www.cimne.com  -

T.(+34) 93 401 56 96 skype: *rougered4*



<http://www.cimne.com/>

<https://www.facebook.com/cimne> <http://blog.cimne.com/>
<http://vimeo.com/cimne> <http://www.youtube.com/user/CIMNEvideos>
<http://www.linkedin.com/company/cimne> <https://twitter.com/cimne>

Les dades personals contingudes en aquest missatge són tractades amb la
finalitat de mantenir el contacte professional entre CIMNE i voste. Podra
exercir els drets d'accés, rectificació, cancel·lació i oposició,
dirigint-se a cimne@cimne.upc.edu. La utilització de la seva adreça de
correu electronic per part de CIMNE queda subjecte a les disposicions de la
Llei 34/2002, de Serveis de la Societat de la Informació i el Comerç
Electronic.

 Imprimiu aquest missatge, només si és estrictament necessari.
<http://www.cimne.com/>

manuelmessmer commented:
Yes, it is also asymmetric: https://github.com/KratosMultiphysics/Kratos/pull/5938#issuecomment-602038978

philbucher commented:
> Switching to LU will solve the problem with the failing test here. But the question why the stiffness matrix is not symmetric and if that is acceptable is still open.

ping @manuelmessmer

manuelmessmer commented:
Sry for the late response. The slight asymmetry seems to come from the element itself. However, this is more sever in my case, because I subtract two stiffness matrices from each other. So I guess, this is not a big problem for other applications. 
I discussed this with @RiccardoRossi. He suggested to have an option (by default enabled) that makes the matrices "artificially" symmetric: K = 0.5 * (K + K^T). What do you think?

philbucher commented:
Where would this option be?

manuelmessmer commented:
Here in this solver.

manuelmessmer commented:
I would add the line K = 0.5*(K + K^T) just before the eigenproblem is solved.

philbucher commented:
Ok dor me

manuelmessmer commented:
I shall push the changes to an extra branch, right?

philbucher commented:
yes, based of master

oberbichler commented:
> Maybe I got this wrong, what does it mean that the stiffness matrix of the shell elements are asymmetric until they are converged?

sorry... missed the question...

AFAIK the only reason for an asymmetrical stiffness matrix is a not 100% clean derivation. In some cases this is for simplification purposes. This does not affect the solution of a structural problem but obviously plays a role when determine singularities. I do not know how much a symmetrization influences the result.
